consensus/hotstuff/consumer.go

```diff
@@ -180,3 +194,42 @@ type QCCreatedConsumer interface {
 	// and must handle repetition of the same events (with some processing overhead).
 	OnQcConstructedFromVotes(*flow.QuorumCertificate)
 }
+
+// TimeoutCollectorConsumer consumes outbound notifications produced by HotStuff and its components.
+// Notifications are consensus-internal state changes which are potentially relevant to
+// the larger node in which HotStuff is running. The notifications are emitted
+// in the order in which the HotStuff algorithm makes the respective steps.
+//
+// Implementations must:
+//   * be concurrency safe
+//   * be non-blocking
+//   * handle repetition of the same events (with some processing overhead).
+type TimeoutCollectorConsumer interface {
+	// OnTcConstructedFromTimeouts notifications are produced by the TimeoutProcessor
+	// component, whenever it constructs a TC from timeouts.
+	// Prerequisites:
+	// Implementation must be concurrency safe; Non-blocking;
+	// and must handle repetition of the same events (with some processing overhead).
+	OnTcConstructedFromTimeouts(certificate *flow.TimeoutCertificate)
+
+	// OnPartialTcCreated notifications are produced by the TimeoutProcessor
+	// component, whenever it aggregates enough timeouts(1/3+1) to claim that TC eventually will be constructed
+	// Prerequisites:
+	// Implementation must be concurrency safe; Non-blocking;
+	// and must handle repetition of the same events (with some processing overhead).
+	OnPartialTcCreated(view uint64)
+
+	// OnNewQcDiscovered notifications are produced by the TimeoutCollector
+	// component, whenever it discovers new QC included in timeout object.
```

Comment:
```suggestion
	// OnNewQcDiscovered notifications are produced by a TimeoutCollector for view v,
	// whenever it discovers a new QC embedded in a TimeoutObject for view v, iff the
	// the new QC has a view greater than any other QC observed in a TimeoutObject for view v.
	// 
```
This is my understanding of how this notification would behave, since each `TimeoutCollector` (view-scoped) will make independent assessments of their local newest QC.
* I'm assuming here that instances of `TimeoutCollectorConsumer` are intended to be global (not view-scoped), based on how the consumer for `QCCreatedConsumer` works

❓  If this is right, I'm wondering if we have an intended use for these notifications currently? They don't seem that useful to me right now, since we will not have QCs ordered by view (even less strict guarantees on ordering than the caveat Alex mentioned, I think), and we will have duplicate QCs in the notifications.

Reply:
## Answer to question 1: 
> `TimeoutCollectorConsumer` is intended to be global (not view-scoped)

Correct.

## My thoughts regarding question 2:
> Do we have an intended use for these notifications?

From my perspective, we strongly desire these notifications. 

### Context:
* `TimeoutObject`s contain QCs and TCs. In the HotStuff liveness proofs, we require that these QCs and TCs -- specifically the newest TC and QC -- are visible to the HotStuff state machine (in our implementation the `EventHander`) to perform the necessary view changes. 
* The `TimeoutObject`s arrive at the [`TimeoutAggregator`](https://github.com/onflow/flow-go/blob/90b4f79a88d5bcfdbfa11f07d80e7144df59cf40/consensus/hotstuff/timeout_aggregator.go), which is its own component, which runs concurrently to the `EventLoop`. 
   *  I think from a theoretical perspective, liveness is guaranteed if the newest QCs and TCs are transmitted to the `EventHandler` as part of a TC or a `partialTC`. 
   * However, waiting for a TC or a `partialTC` could have a large penalty in practise: 
       * Consider the case where node `Bob` is primary for view `100`.  View 99 has timed out. But Bob is behind still stuck in view 98. Maybe he just recovered from a crash and missed most of the TOs for view 99. 
       * Now, when the first node in view 100 times out and sends a TO, the information contained therein would allow Bob to transition to view 100, quickly construct his block and broadcast it. Hopefully in time that a supermajority of nodes still get the proposal, vote for it and the view completes successfully.
       * However,  if the `TimeoutAggregator` does not expose new information to the `EventHandler` asap, view 100 will likely time out. By the time Bob's `TimeoutAggregator` constructs a `partialTC`, too many nodes have timed out so no QC can be constructed for Bob's proposal anymore. 
* In a nutshell, `TimeoutAggregator` receives QCs and TCs that the `EventHandler` needs to transition to the newest view and contribute to consensus. Waiting for a `partialTC` to propagate this information adds latency. The extra latency can result extra views timing out. I think this might be a relatively likely scenario when the committee is operating close to its timeout threshold. 
* Hence, we desire a _fast_ channel over which the `TimeoutAggregator` can push the _newest_ QCs or TCs to the `EventHander`. This channel is implemented via `OnNewQcDiscovered` and `OnTcConstructedFromTimeouts`

### Completeness: newest QC and TC is eventually delivered and _never_ lost
Reasoning:
* based on the context above, the newest QC and TC must reach `EventHander` (skipping QCs or TCs with lower view is fine)
* It is straight to show that the current implementation will _always_ deliver the newest QC and TC to the `EventHander`. QCs and TCs for older views might be interjected rarely; however the `EventHander` can just drop older QCs or TCs by inspecting their view alone. 

### Responsiveness: newest QC and TC is _quickly_ delivered with negligible delay 
Reasoning:
* Our logic only only generates a `OnNewQcDiscovered ` notification if the QC's view is strictly larger than any QC that was sent before: https://github.com/onflow/flow-go/blob/90b4f79a88d5bcfdbfa11f07d80e7144df59cf40/consensus/hotstuff/timeoutcollector/timeout_collector.go#L81-L87 In other words, for each view, at most a single `OnNewQcDiscovered ` is sent. Hence, it is extremely unlikely for the notification channel to become congested with notifications about old QCs while the newest QC is delayed. 
* similar argument applies to TC

Summary:
We have reasoned that `OnNewQcDiscovered` and `OnTcConstructedFromTimeouts` are desired for the `TimeoutAggregator` to quickly push new QCs and TCs to the `EventHander` to improve responsiveness of the consensus protocol. Furthermore, we have sketched proofs that the current implementation completely delivers the necessary information without noteworthy latency.

Reply:
## Follow-up comment
**Caution:** I think the current implementation of `OnPartialTcCreated` is potentially insufficient to guarantee liveness: https://github.com/onflow/flow-go/blob/90b4f79a88d5bcfdbfa11f07d80e7144df59cf40/consensus/hotstuff/consumer.go#L220

@durkmurder

Reply:
## Follow-up comment 2

We also have to think about how we handle a QC for a block that we haven't received yet.

Reply:
Thanks for the explanation Alex, it makes perfect sense now 👌 

> > We also have to think about how we handle a QC for a block that we haven't received yet.
> 
> Can you elaborate more on this topic?

Here is my understanding:
* Previously, the only way for a new QC to enter HotStuff was as part of a `BlockProposal`. Since The compliance layer guarantees that parent blocks are processed before children, we could be certain that we know about the block pointed to by any QC which is processed by HotStuff. 
* However `TimeoutObject`s are processed separately without a similar ordering constraint, so we may process a `TimeoutObject` containing a QC for block `B`, before we process block `B`
  * This would cause a (currently unexpected) error [when adding the QC to `Forks`](https://github.com/onflow/flow-go/blob/1e0a30e7a9118aad96a706639e37ec955f562542/consensus/hotstuff/forks/forkchoice/newest.go#L93)
* This is similar to the [discussion about processing blocks containing TCs on the Implementation doc](https://www.notion.so/Active-Pacemaker-Implementation-Draft-f654eaa990954508908304ba03428a24?d=3029169ae9b4468396226451bd63d02c#3888981b6b5646e6aed0a2a7df3a5bb4)

I think ultimately the problem is, we can process QCs for block `B` without having processed `B`. But this complicates our proposal and voting logic, because:
* we would like to build on top of our highest known QC, but we can't propose a block on top of a block `B` if we haven't validated the payload for block `B`
* we can't vote for a proposal building on top of block `B` if we haven't validated the payload for block `B` yet

I see two options:
1. we can avoid processing QCs (or messages containing QCs) referencing an unknown block, eg. by queueing messages at the 
2. we can optimistically process QCs, but voting and proposing logic (and internal HotStuff state) needs to be aware of the possibility of a QC for an unknown block

To me, Option 1 is preferable because it limits more edge cases within HotStuff and is consistent with the approach taken for TCs (linked discussion) of optimizing for happy-path of voting or proposing.

Reply:
> Lemma: Bracha broadcast is need to ensure that we produce a TC as fast as possible by immediately timing out current view, this ensures lowest possible time to enter round `r` within `2*Δ` after GST. 

This is precisely the argument we need to look at. Thanks.

I think the current implementation is _not_ sufficient to guarantee that lowest possible time to enter round `r` within `2*Δ` after GST. Reasoning:
* consider the following situation: 
   * a replica is collecting enough timeouts for view 100 such that `TimeoutAggregator` emits a `OnPartialTcCreated` notification
   * however, the replica's `EventHander` is still in view 99 (it hasn't seen the necessary evidence to transition to view 100 yet)
* `OnPartialTcCreated` for view 100 arrives at the `EventHander`, the event hander calls into `SafetyRules`  https://github.com/onflow/flow-go/blob/90b4f79a88d5bcfdbfa11f07d80e7144df59cf40/consensus/hotstuff/safetyrules/safety_rules.go#L123 
   * But neither the QC nor the TC allow for switching to view 100.
   * Hence, `SafetyRules` will not generate a `TimeoutMessage` and   `EventHander` will not enter the view in time.
  
I would conclude that our implementation does not guarantee validity of the Lemma.

Reply:
> > We also have to think about how we handle a QC for a block that we haven't received yet.
>
> Can you elaborate more on this topic? I am not quite sure what do you exactly mean.

exactly what [@jordanschalm said](https://github.com/onflow/flow-go/pull/2620#discussion_r902007880).

>I see two options:
> 1. we can avoid processing QCs (or messages containing QCs) referencing an unknown block, eg. by queueing messages before they are processed by `TimeoutAggregator`.
> 2. we can optimistically process QCs, but voting and proposing logic (and internal HotStuff state) needs to be aware of the possibility of a QC for an unknown block.

I think both options generate undesired edge cases. My gut feeling is that option 2 might be better. With option 2, we might halt the generation of TCs locally, if only a few `TimeoutObject`s refer to unknown blocks (since they are cached and don't reach the `TimeoutAggregator`).

I still have the vague hope that we can adjusting the `EventHander` might be less complex. The node can't propose or vote if it doesn't have sufficiently new blocks, but it can at least participate in the TC production. Eventually a node that knows the highest block will become primary (existence of a QC implies that a supermajority of nodes know the block) and propose a proper extension. 
Cycling through leaders requires an intact timeout mechanism; hence my concern with withholding messages from the `TimeoutAggregator`

Reply:
Result of discussion:
* we'll implement option 2 (QCs and TCs are forwarded to EventHander, which progresses through views but might not vote if parent block is unknown).